File: src/types/endpointProtection.ts
// Endpoint protection GPO types

export type EndpointGpoType =
  | "AuditBaseline"
  | "AuditEnhanced"
  | "DcAuditEssential"
  | "DcAuditComprehensive"
  | "DefenderProtection";

export type EndpointGpoCategory = "audit" | "dc-audit" | "defender";

export type EndpointGpoLinkScope = "per-tier" | "dc-only" | "domain-wide";

// Status of a tier's link for per-tier GPOs
export interface TierLinkStatus {
  tier: string;
  linked: boolean;
  linkEnabled: boolean;
}

// Status of an endpoint protection GPO
export interface EndpointGpoStatus {
  gpoType: string;
  name: string;
  description: string;
  exists: boolean;
  linked: boolean;
  linkTarget: string;
  linkScope: string;
  created: string | null;
  modified: string | null;
  tierStatus: TierLinkStatus[] | null;
}

// Result from endpoint GPO configuration
export interface EndpointGpoConfigResult {
  success: boolean;
  gpoType: string;
  gpoName: string;
  created: boolean;
  linked: boolean;
  configured: boolean;
  errors: string[];
  warnings: string[];
}

// GPO configuration metadata
export interface EndpointGpoConfig {
  type: EndpointGpoType;
  name: string;
  description: string;
  category: EndpointGpoCategory;
  linkScope: EndpointGpoLinkScope;
  features: string[];
}

// Configuration metadata for all endpoint protection GPOs
export const ENDPOINT_GPO_CONFIGS: EndpointGpoConfig[] = [
  {
    type: "AuditBaseline",
    name: "SEC-Audit-Baseline",
    description: "Microsoft recommended baseline audit policies",
    category: "audit",
    linkScope: "per-tier",
    features: [
      "Account Logon: Success/Failure",
      "Account Management: Success/Failure",
      "Logon/Logoff: Success/Failure",
      "Object Access: Failure",
      "Policy Change: Success/Failure",
      "Privilege Use: Failure",
      "System: Success/Failure",
    ],
  },
  {
    type: "AuditEnhanced",
    name: "SEC-Audit-Enhanced",
    description: "ACSC/NSA hardened audit policies with PowerShell logging",
    category: "audit",
    linkScope: "per-tier",
    features: [
      "All Baseline settings",
      "Process Creation with Command Line",
      "PowerShell Script Block Logging",
      "PowerShell Module Logging",
      "PowerShell Transcription",
      "Detailed File Share Access",
      "Registry Auditing",
      "Removable Storage",
      "DPAPI Activity",
    ],
  },
  {
    type: "DcAuditEssential",
    name: "SEC-DC-Audit-Essential",
    description: "Essential security audit policies for Domain Controllers",
    category: "dc-audit",
    linkScope: "dc-only",
    features: [
      "Directory Service Access",
      "Directory Service Changes",
      "Directory Service Replication",
      "Kerberos Account Logon",
      "Credential Validation",
      "Security Group Management",
      "User Account Management",
    ],
  },
  {
    type: "DcAuditComprehensive",
    name: "SEC-DC-Audit-Comprehensive",
    description: "Comprehensive forensic audit policies for Domain Controllers",
    category: "dc-audit",
    linkScope: "dc-only",
    features: [
      "All Essential settings",
      "Kerberos Authentication Service",
      "Kerberos Service Ticket Operations",
      "LDAP Interface Events",
      "Detailed Replication",
      "SAM Operations",
      "Process Creation with Command Line",
      "Certification Services",
    ],
  },
  {
    type: "DefenderProtection",
    name: "SEC-Defender-Protection",
    description: "Microsoft Defender Antivirus balanced protection settings",
    category: "defender",
    linkScope: "domain-wide",
    features: [
      "Real-time Protection",
      "Cloud-delivered Protection",
      "Automatic Sample Submission",
      "PUA Protection (Block)",
      "Behavior Monitoring",
      "Email Scanning",
      "Removable Drive Scanning",
      "Archive Scanning",
    ],
  },
];

// Get config by type
export function getEndpointGpoConfig(type: EndpointGpoType): EndpointGpoConfig | undefined {
  return ENDPOINT_GPO_CONFIGS.find((c) => c.type === type);
}

// Get configs by category
export function getEndpointGposByCategory(category: EndpointGpoCategory): EndpointGpoConfig[] {
  return ENDPOINT_GPO_CONFIGS.filter((c) => c.category === category);
}

File: src-tauri/src/commands/endpoint_protection.rs
//! Tauri commands for endpoint protection GPO management

use crate::domain::Tier;
use crate::infrastructure::endpoint_protection::{
    self, EndpointGpoConfigResult, EndpointGpoStatus, EndpointGpoType,
};
use crate::infrastructure::get_domain_dn;

/// Get status of all endpoint protection GPOs
#[tauri::command]
pub async fn get_endpoint_protection_status() -> Result<Vec<EndpointGpoStatus>, String> {
    let domain_dn = get_domain_dn().map_err(|e| e.to_string())?;
    endpoint_protection::get_all_endpoint_gpo_status(&domain_dn).map_err(|e| e.to_string())
}

/// Configure an endpoint protection GPO
#[tauri::command]
pub async fn configure_endpoint_gpo(
    gpo_type: String,
    tier: Option<String>,
) -> Result<EndpointGpoConfigResult, String> {
    let domain_dn = get_domain_dn().map_err(|e| e.to_string())?;

    let tier_enum = tier.as_ref().and_then(|t| match t.as_str() {
        "Tier0" => Some(Tier::Tier0),
        "Tier1" => Some(Tier::Tier1),
        "Tier2" => Some(Tier::Tier2),
        _ => None,
    });

    match gpo_type.as_str() {
        "AuditBaseline" => {
            if let Some(t) = tier_enum {
                endpoint_protection::configure_audit_baseline_gpo(t, &domain_dn)
                    .map_err(|e| e.to_string())
            } else {
                Err("Tier is required for AuditBaseline GPO".to_string())
            }
        }
        "AuditEnhanced" => {
            if let Some(t) = tier_enum {
                endpoint_protection::configure_audit_enhanced_gpo(t, &domain_dn)
                    .map_err(|e| e.to_string())
            } else {
                Err("Tier is required for AuditEnhanced GPO".to_string())
            }
        }
        "DcAuditEssential" => {
            endpoint_protection::configure_dc_audit_essential_gpo(&domain_dn)
                .map_err(|e| e.to_string())
        }
        "DcAuditComprehensive" => {
            endpoint_protection::configure_dc_audit_comprehensive_gpo(&domain_dn)
                .map_err(|e| e.to_string())
        }
        "DefenderProtection" => {
            endpoint_protection::configure_defender_gpo(&domain_dn).map_err(|e| e.to_string())
        }
        _ => Err(format!("Unknown GPO type: {}", gpo_type)),
    }
}

/// Configure all endpoint protection GPOs
#[tauri::command]
pub async fn configure_all_endpoint_gpos() -> Result<Vec<EndpointGpoConfigResult>, String> {
    let domain_dn = get_domain_dn().map_err(|e| e.to_string())?;
    let mut results = Vec::new();

    // Configure per-tier audit GPOs
    for tier in Tier::all() {
        // Baseline audit
        match endpoint_protection::configure_audit_baseline_gpo(*tier, &domain_dn) {
            Ok(result) => results.push(result),
            Err(e) => {
                let mut result =
                    EndpointGpoConfigResult::new("AuditBaseline", &format!("SEC-{}-Audit-Baseline", tier));
                result.add_error(e.to_string());
                results.push(result);
            }
        }

        // Enhanced audit
        match endpoint_protection::configure_audit_enhanced_gpo(*tier, &domain_dn) {
            Ok(result) => results.push(result),
            Err(e) => {
                let mut result =
                    EndpointGpoConfigResult::new("AuditEnhanced", &format!("SEC-{}-Audit-Enhanced", tier));
                result.add_error(e.to_string());
                results.push(result);
            }
        }
    }

    // Configure DC audit GPOs
    match endpoint_protection::configure_dc_audit_essential_gpo(&domain_dn) {
        Ok(result) => results.push(result),
        Err(e) => {
            let mut result = EndpointGpoConfigResult::new("DcAuditEssential", "SEC-DC-Audit-Essential");
            result.add_error(e.to_string());
            results.push(result);
        }
    }

    match endpoint_protection::configure_dc_audit_comprehensive_gpo(&domain_dn) {
        Ok(result) => results.push(result),
        Err(e) => {
            let mut result =
                EndpointGpoConfigResult::new("DcAuditComprehensive", "SEC-DC-Audit-Comprehensive");
            result.add_error(e.to_string());
            results.push(result);
        }
    }

    // Configure Defender GPO
    match endpoint_protection::configure_defender_gpo(&domain_dn) {
        Ok(result) => results.push(result),
        Err(e) => {
            let mut result = EndpointGpoConfigResult::new("DefenderProtection", "SEC-Defender-Protection");
            result.add_error(e.to_string());
            results.push(result);
        }
    }

    Ok(results)
}

/// Delete an endpoint protection GPO
#[tauri::command]
pub async fn delete_endpoint_gpo_cmd(
    gpo_type: String,
    tier: Option<String>,
) -> Result<(), String> {
    let tier_enum = tier.as_ref().and_then(|t| match t.as_str() {
        "Tier0" => Some(Tier::Tier0),
        "Tier1" => Some(Tier::Tier1),
        "Tier2" => Some(Tier::Tier2),
        _ => None,
    });

    let gpo_type_enum = match gpo_type.as_str() {
        "AuditBaseline" => EndpointGpoType::AuditBaseline,
        "AuditEnhanced" => EndpointGpoType::AuditEnhanced,
        "DcAuditEssential" => EndpointGpoType::DcAuditEssential,
        "DcAuditComprehensive" => EndpointGpoType::DcAuditComprehensive,
        "DefenderProtection" => EndpointGpoType::DefenderProtection,
        _ => return Err(format!("Unknown GPO type: {}", gpo_type)),
    };

    endpoint_protection::delete_endpoint_gpo(gpo_type_enum, tier_enum).map_err(|e| e.to_string())
}
